plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/manager/VmwareStorageManagerImpl.java

```diff
@@ -1464,4 +1448,55 @@ private String deleteDir(String dir) {
     private static String getVolumeRelativeDirInSecStroage(long volumeId) {
         return "volumes/" + volumeId;
     }
+
+    /**
+     * This method is to check if a given TaskInfo Object is valid( and has name and entity name assigned).It return true if TaskInfo Object is valid and false otherwise.
+     *
+     * @param TaskInfo
+     *        info
+     * @return boolean(true or false)
+     **/
+    private boolean isvalidTaskInfoObj(TaskInfo info){
+        return !(info == null || info.getEntityName() == null || info.getName() == null);
+    }
+    
+    /**
+     * This method waits for tasks running on vm to complete
+     *
+     * @param vmName        Name of the vm
+     * @param taskName      Name of the task
+     * @param context       Task context object
+     * 
+     * @return boolean(true or false)   True if it waited for tasks to finish and false when there are no tasks running on the Vm.
+     **/
+     
+    private boolean waitForRunningTaskOnVM(String vmName, String taskName, VmwareContext context) throws Exception {
+        try {
+            ManagedObjectReference taskmgr = context.getServiceContent().getTaskManager();
+            List<ManagedObjectReference> tasks = context.getVimClient().getDynamicProperty(taskmgr, "recentTask");
+
+            for (ManagedObjectReference taskMor : tasks) {
+                TaskInfo info = (TaskInfo)(context.getVimClient().getDynamicProperty(taskMor, "info"));
+
+                if (!isvalidTaskInfoObj(info)) {
+                    continue;
+                }
+
+                if (info.getEntityName().equals(vmName) && info.getName().equalsIgnoreCase(taskName)) {
+                    if (!(info.getState().equals(TaskInfoState.SUCCESS) || info.getState().equals(TaskInfoState.ERROR))) {
+                        s_logger.debug("There is already a task: " + taskName + " running for VM: " + vmName + ", wait for it");
+                        // wait if there is already similar VM task running
+                        context.getVimClient().waitForTask(taskMor);
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        } catch (Exception e) {
+            String msg = "Failed to check running task: " + taskName + " for vm: " + vmName + " due to " + e.getMessage();
+            s_logger.error(msg, e);
+            throw new Exception(msg);
```

Comment:
`Exception` is a generic type, and, generally, should not be thrown.  I would recommend `IllegalStateException` here.